File: dp_labs_and_solutions/dp_labs_and_solutions/dp_cpp_labs_and_solutions/singleton/abstract_factory/lab/garden.cpp
// garden.cpp

#include "dp_exception.h"
#include "lettuce.h"
#include "garden.h"
#include "conventional_garden.h"
#include "organic_garden.h"
#include "tomato.h"
#include "conventional_lettuce.h"
#include "organic_lettuce.h"

garden* garden::the_instance(nullptr);

garden::garden(lettuce* a_lettuce_prototype)
{
    set_lettuce_prototype(a_lettuce_prototype);
}

garden::~garden()
{
    delete my_lettuce;
    my_lettuce = nullptr;
}

void garden::set_lettuce_prototype(lettuce* a_lettuce_prototype)
{
    if (a_lettuce_prototype == nullptr)
    {
        dp_exception("a_lettuce_prototype is NULL!");
    }
    my_lettuce = a_lettuce_prototype->clone();
}

garden& garden::get_instance()
{
	if (the_instance == nullptr)
	{
		if (1)
		{
			the_instance = new conventional_garden(new conventional_lettuce(2, 1));
		}
		else
		{
			the_instance = new organic_garden(new organic_lettuce(2, 1));
		}
	}
	return *the_instance;
}
lettuce * garden::make_lettuce() const
{
	return my_lettuce;
}

File: dp_labs_and_solutions/dp_labs_and_solutions/dp_matlab_labs_and_solutions/singleton/abstract_factory/solution/Garden.m
% Garden.m

classdef (Abstract) Garden < handle
% Singleton:

	methods (Static)
		function result = getInstance
			persistent theInstance;
			if isempty(theInstance)
				if true
					theInstance = ConventionalGarden(ConventionalLettuce(1, 2));
				else
					theInstance = OrganicGarden(OrganicLettuce(1, 2));
				end
			end
			result = theInstance;
		end
	end

% non-Singleton:

	properties
		LettucePrototype % Lettuce
	end

	methods (Access = protected)
		% constructor:
		function theGarden = Garden(aLettucePrototype)
			theGarden.LettucePrototype = aLettucePrototype;
		end
	end

	methods
		function set.LettucePrototype(theGarden, aLettucePrototype)
			if isempty(aLettucePrototype)
				error('A Garden must have a valid Lettuce prototype.');
			end

			theGarden.LettucePrototype = aLettucePrototype.clone;
		end

		function result = makeLettuce(theGarden)
			result = theGarden.LettucePrototype.clone;
		end
	end

	methods (Abstract)
		makeTomato(theGarden)
	end
end
